rush/rush/src/actions/GenerateAction.ts

```diff
@@ -176,26 +277,67 @@ export default class GenerateAction extends CommandLineAction {
+    //    necessary dependencies. This will happen if someone is adding a new rush dependency,
+    //    or if someone is adding a dependency which already exists in another project or exists
+    //    as an indirect dependency at the root of the common/node_modules
+    const shouldClean: boolean =
+      GenerateAction._shouldDeleteNodeModules(this._rushConfiguration, tempModules);
 
-    // 4. Construct common\package.json and common\temp_modules
-    GenerateAction._createCommonTempModulesAndPackageJson(this._rushConfiguration);
+    if (isLazy) {
+      console.log(colors.green(
+        `${os.EOL}Rush is running in "--lazy" mode. ` +
+        `You will need to run "rush generate" before committing.`));
+    } else if (!shouldClean) {
+      console.log(colors.green(
+        `${os.EOL}Rush found all dependencies in the shrinkwrap. Skipping some expensive steps!`));
+    } else {
+      console.log(colors.yellow(`${os.EOL}The shrinkwrap file was missing one or more dependencies. ` +
+        `Rush must run in "--clean" mode, which deletes and replaces the node_modules folder.${os.EOL}` +
+        `This operation may take some time...`));
+    }
+
+    // 4. Delete "common\node_modules" or "common\node_modules\rush-*"
+    if (isLazy || !shouldClean) {
+      GenerateAction._deleteCommonRushModules(this._rushConfiguration);
+    } else {
+      GenerateAction._deleteCommonNodeModules(this._rushConfiguration);
+    }
 
-    // 5. Make sure the NPM tool is set up properly.  Usually "rush install" should have
+    if (isLazy || shouldClean) {
+      // 5. Delete the previous npm-shrinkwrap.json
+      GenerateAction._deleteShrinkwrapFile(this._rushConfiguration);
+    }
+
+    // 6. Make sure the NPM tool is set up properly.  Usually "rush install" should have
     //    already done this, but not if they just cloned the repo
     console.log('');
     InstallAction.ensureLocalNpmTool(this._rushConfiguration, false);
 
-    // 6. Run "npm install" and "npm shrinkwrap"
+    // 7. Run "npm install" and "npm shrinkwrap"
+    // (always run, installing based on old shrinkwrap if temp_modules has changed but no need to create new shrinkwrap
     GenerateAction._runNpmInstall(this._rushConfiguration);
-    GenerateAction._runNpmShrinkWrap(this._rushConfiguration, this._lazyParameter.value);
+
+    // 8. If we are doing a "fast" generate, then we need to make sure anything that was removed gets deleted from n_m
+    if (!shouldClean) {
+      GenerateAction._deleteShrinkwrapFile(this._rushConfiguration);
+      GenerateAction._runNpmPrune(this._rushConfiguration);
+      GenerateAction._runNpmDedupe(this._rushConfiguration);
+    }
+
+    if (isLazy) {
+      // Don't shrinkwrap the --lazy generate to remind the user they should run a regular generate
+      console.log(os.EOL + colors.bold('(Skipping "npm shrinkwrap")') + os.EOL);
+    } else {
+      GenerateAction._runNpmShrinkWrap(this._rushConfiguration);
+    }
```

Comment:
We also realized that after these changes, the temp_modules don't need to be in Git at all any more.  We said the user interface should be like this:

- **rush install**:  Always regenerates the temp_modules before installing.  However, if shrinkwrap is out of date, then it fails (after zero seconds!), telling the person to run "rush upgrade".  (It's not automatic, because "upgrade" takes long and is a complicated PR approval.)

- **rush upgrade**:  Always does a full "generate" of everything, even if shrinkwrap is technically valid.  Supports the `--lazy` option, which is really a tool to help people who need to do multiple attempts as part of a full upgrade of everything.